File: SWEA/D2/12712. 파리퇴치3/파리퇴치3.py
T = int(input())


for test_case in range(1, T + 1):
    n, m = map(int, input().split())
    arr = [[0] * 3 * n for _ in range(3 * n)]
    for i in range(n, 2 * n):
        temp = list(map(int, input().split()))
        for j in range(n, 2 * n):
            arr[i][j] = temp[j - n]

    max = 0
    for i in range(n, 2 * n):
        for j in range(n, 2 * n):
            sum = 0
            for h in range(-m + 1, m):
                sum += arr[i][j + h]
                sum += arr[i + h][j]
            sum -= arr[i][j]
            if sum > max:
                max = sum

            sum = 0
            for h in range(-m + 1, m):
                sum += arr[i + h][j + h]
                sum += arr[i + h][j - h]
            sum -= arr[i][j]
            if sum > max:
                max = sum

    print("#{} {}" .format(test_case, max))

File: SWEA/D2/12712. 파리퇴치3/파리퇴치3.java
import java.util.*;

public class Solution {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        for(int test_case = 1; test_case <= T; test_case++) {
            int n = sc.nextInt();
            int m = sc.nextInt();
            int[][] arr = new int[n][n];
            for(int i = 0; i<n; i++) {
                for(int j = 0; j<n; j++) {
                    arr[i][j] = sc.nextInt();
                }
            }

            int max = 0;
            for(int i = 0; i<n; i++) {
                for(int j = 0; j<n; j++) {
                    int sum = arr[i][j];
                    for(int h = 1; h<m; h++) {
                        if(j + h < n) {
                            sum += arr[i][j + h];
                        }
                        if(j - h >= 0) {
                            sum += arr[i][j - h];
                        }
                        if(i + h < n) {
                            sum += arr[i + h][j];
                        }
                        if(i - h >= 0) {
                            sum += arr[i - h][j];
                        }
                    }
                    if(sum > max)
                        max = sum;
                }
            }

            for(int i = 0; i<n; i++) {
                for(int j = 0; j<n; j++) {
                    int sum = arr[i][j];
                    for(int h = 1; h<m; h++) {
                        if(j + h < n && i + h < n) {
                            sum += arr[i + h][j + h];
                        }
                        if(j - h >= 0 && i - h >= 0) {
                            sum += arr[i - h][j - h];
                        }
                        if(i + h < n && j - h >= 0) {
                            sum += arr[i + h][j - h];
                        }
                        if(i - h >= 0 && j + h < n) {
                            sum += arr[i - h][j + h];
                        }
                    }
                    if(sum > max)
                        max = sum;
                }
            }

            System.out.print("#" + test_case + " ");
            System.out.println(max);
        }
    }
}